lib/pidtree.js

```diff
@@ -0,0 +1,82 @@
+'use strict';
+
+var getAll = require('./get');
+
+/**
+ * Get the list of children pids of the given pid.
+ * @param  {Number|String} pid A pid.
+ * @param  {Object} [options] Optional options object.
+ * @param  {Boolean} [options.root=false] Include the provided pid in the list.
+ * @param  {pidCallback} callback Called when the list is ready.
+ */
+function list(pid, options, callback) {
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  if (typeof options !== 'object') {
+    options = {};
+  }
+
+  pid = parseInt(pid, 10);
+  if (isNaN(pid) || pid < -1) {
+    callback(new TypeError('The pid provided is invalid'));
+    return;
+  }
+
+  getAll(function(err, list) {
+    if (err) {
+      callback(err);
+      return;
+    }
+    if (pid === -1) {
+      for (var i = 0; i < list.length; i++) {
+        list[i] = {ppid: list[i][0], pid: list[i][1]};
+      }
+      callback(null, list);
+      return;
+    }
+
+    var root;
+    for (var l = 0; l < list.length; l++) {
+      if (list[l][1] === pid) {
+        root = {ppid: list[l][0], pid: list[l][1]};
+        break;
+      }
+    }
+    if (!root) {
+      callback(new Error('No maching pid found'));
+      return;
+    }
+
+    // Build the graph
+    var tree = {};
+    while (list.length > 0) {
+      var e = list.pop();
+      if (tree[e[0]]) {
+        tree[e[0]].push(e[1]);
+      } else {
+        tree[e[0]] = [e[1]];
+      }
+    }
+    // Simplified BFS alike tree visit
+    var idx = 0;
+    var pids = [root];
+    while (idx < pids.length) {
+      var curpid = pids[idx++].pid;
+      if (!tree[curpid]) continue;
+      var len = tree[curpid].length;
+      for (var j = 0; j < len; j++) {
+        pids.push({ppid: curpid, pid: tree[curpid][j]});
+      }
+      delete tree[curpid];
+    }
```

Comment:
Can't those two steps be made in one pass?

Reply:
The first step builds the [adjacency list](https://en.wikipedia.org/wiki/Adjacency_list)
Then after that I start visiting the tree from the root node using the a [BFS](https://en.wikipedia.org/wiki/Breadth-first_search).

BUT I can build the vector of parents instead of an adjacency list.
But then I should do a leaf to parent visit. Trickier to write but should be faster.

Reply:
This is the implementation using the vector of parents.
Maybe is better to just keep the old one, should be easier to maintain.
This should be a bit more memory efficient since in general pid trees are more wider than deeper but practically there isn't such a big difference.
```js
// Build the HashMap of parents (pid -> ppid)
    var info = {};
    var pids = [];
    while (list.length > 0) {
      var e = list.pop();
      info[e[1]] = {parent: e[0]};
      pids.push(e[1]);
    }

    // For each pid in the system:
    //  - Traverse the path from leaf to root until:
    //    - the info of whether or not the current pid is grandchildren of the
    //      PID the user is interested in is already known.
    //    - OR
    //    - the current pid is a root.
    //  - Mark each node in the path from leaf to root with the info of
    //    whether or not they are grandchildren of the PID the user is
    //    interested.
    while (pids.length > 0) {
      var last = pids.pop();
      var path = [];
      var status = false;
      // eslint-disable-next-line no-constant-condition
      while (true) {
        if (!info[last]) {
          info[last] = {};
        }
        var data = info[last];
        // Insert the pid in the path.
        path.push(last);
        if (last === PID || data.ancestor === true) {
          // This pid is the one the user is interested in or is one of its
          // grandchildren
          status = true;
          break;
        }
        if (data.ancestor === false || data.parent === undefined) {
          // This pid in not a grandchildren of the PID the user is interested in
          // or this is a root.
          break;
        }
        last = data.parent;
      }
      // Mark all the PIDS in the path traversed with the info of whether or not
      // they are grandchildren of the PID the user is interested.
      if (info[path[path.length - 1]] === undefined) {
        info[path[path.length - 1]] = {};
      }
      while (path.length > 0) {
        info[path.pop()].ancestor = status;
      }
    }

    // Do a linear scan to get the list of pids that are children or
    // grandchildren of the PID the user is interested.
    // eslint-disable-next-line guard-for-in
    for (var pid in info) {
      if (info[pid].ancestor) {
        pids.push(
          options.advanced
            ? {ppid: info[pid].parent, pid: Number(pid)}
            : Number(pid)
        );
      }
      delete info[pid];
    }

    if (!options.root) {
      pids.shift(); // Remove root
    }
```